bindings/cs/Protean-cs/Protean.Net/BinaryWriter.cs

```diff
@@ -65,357 +65,561 @@ public static byte[] ToBytes(Variant value, BinaryMode mode)
-
-		protected void Write(DateTime arg)
-		{
-			if (arg == DateTime.MaxValue)
-			{
-				Write(MaxDateTimeMillis);
-			}
-			else
-			{
-				Write((arg.Ticks - Variant.MinDateTime.Ticks) / 10000);
-			}
-		}
-
-		private void Write(byte arg)
-		{
-			m_filter.WriteByte(arg);
-		}
-
-		protected void Write(Int32 arg)
-		{
-			Write(BitConverter.GetBytes(arg));
-		}
-
-		protected void Write(UInt32 arg)
-		{
-			Write(BitConverter.GetBytes(arg));
-		}
-
-		protected void Write(Int64 arg)
-		{
-			Write(BitConverter.GetBytes(arg));
-		}
-
-		protected void Write(UInt64 arg)
-		{
-			Write(BitConverter.GetBytes(arg));
-		}
-
-		protected void Write(float arg)
-		{
-			Write(BitConverter.GetBytes(arg));
-		}
-
-		protected void Write(double arg)
-		{
-			Write(BitConverter.DoubleToInt64Bits(arg));
-		}
-
-		private void Write(byte[] arg)
-		{
-			m_filter.Write(arg, 0, arg.Length);
-		}
-
-		protected void Write(bool arg)
-		{
-			Write((arg ? 1 : 0));
-		}
-
-		protected void Write(byte[] bytes, bool writePadding)
-		{
-			Write(bytes);
-
-			if (writePadding)
-			{
-				int residual = (4 - (bytes.Length % 4)) % 4;
-				for (int i = 0; i < residual; ++i)
-				{
-					Write((byte)0);
-				}
-			}
-		}
-
-	    protected void Write(DataTable arg)
-	    {
+        }
+
+        public static byte[] ToBytes(Variant value)
+        {
+            return ToBytes(value, BinaryMode.Default);
+        }
+
+        public void Write(Variant v)
+        {
+            WriteHeader();
+            WriteVariant(v);
+        }
+
+        public static int GetBytes(Variant value, BinaryMode mode, byte[] buffer, int bufferIndex)
+        {
+            using (MemoryStream ms = new MemoryStream(buffer, bufferIndex, buffer.Length - bufferIndex))
+            {
+                using (BinaryWriter writer = new BinaryWriter(ms, mode))
+                {
+                    writer.Write(value);
+                }
+
+                return (int)ms.Position;
+            }
+        }
+
+        public static int GetByteCount(Variant variant)
+        {
+            return 12 + GetByteCountVariant(variant);
+        }
+
+        private void WriteHeader()
+        {
+            // Create and write the Protean header
+            m_stream.Write(BitConverter.GetBytes(BinaryConstants.PROTEAN_MAGIC), 0, 4);
+            m_stream.Write(BitConverter.GetBytes(BinaryConstants.VERSION_MAJOR << 16 | BinaryConstants.VERSION_MINOR), 0, 4);
+            m_stream.Write(BitConverter.GetBytes((int)m_mode), 0, 4);
+        }
+
+        // This doesn't seem to support Unicode properly
+        private void Write(string arg)
+        {
+            int length = arg.Length;
+            byte[] bytes = new byte[length];
+
+            for (int i = 0; i < length; ++i)
+            {
+                bytes[i] = (byte)arg[i];
+            }
+
+            Write(length);
+            Write(bytes, true);
+        }
+
+        private static int GetByteCountString(string arg)
```

Comment:
If you're going to suffix all of the `GetByteCount` methods, should you update the `Write` methods to have the same? Either way works, but I think we should probably be consistent about it.